### PR TITLE
Rework modules section

### PR DESCRIPTION
**Pull request by yurabakhtin**
**What kind of change does this PR introduce?** (check at least one)

- [x] Feature

**The PR fulfills these requirements:**

- [x] It's submitted to the `develop` branch, _not_ the `master` branch if no hotfix
- [ ] All tests are passing
- [ ] Changelog was modified

**Other information:**
Issue: https://github.com/humhub/team_tasks/issues/169

### CODE REVIEW

**Comment by ArchBlood** [comment]
More of a quick question, would it not be best to move any less files to their own folder?
`<module-id>/resources/less`

File: protected/humhub/modules/admin/views/layouts/module-old.php [hunk 1]
```diff
@@ -0,0 +1,15 @@
+<?php $this->beginContent('@admin/views/layouts/main.php') ?>
```
Comment (luke-) [inline protected/humhub/modules/admin/views/layouts/module-old.php:1]: Can we remove this file?

File: protected/humhub/modules/marketplace/controllers/PurchaseController.php [hunk 2]
```diff
@@ -29,7 +29,7 @@ class PurchaseController extends Controller
     /**
      * @var string
      */
-    public $subLayout = '@admin/views/layouts/module';
+    public $subLayout = '@admin/views/layouts/module-old';
```
Comment (luke-) [inline protected/humhub/modules/marketplace/controllers/PurchaseController.php:32]: Still required?

File: protected/humhub/modules/marketplace/Events.php [hunk 3]
```diff
@@ -34,50 +43,237 @@ public static function onConsoleApplicationInit($event)
...
+                    if (!$module->isProOnly()) {
+                        return false;
+                    }
+                    break;
+                case 'featured':
+                    if (!$module->getOnlineInfo('featured')) {
+                        return false;
+                    }
+                    break;
+                case 'official':
```
Comment (luke-) [inline protected/humhub/modules/marketplace/Events.php:239]: Show this badge when `isThirdParty` is `false`

File: protected/humhub/modules/marketplace/Events.php [hunk 4]
```diff
@@ -34,50 +43,237 @@ public static function onConsoleApplicationInit($event)
...
+                    if (!$module->getOnlineInfo('isCommunity')) {
+                        return false;
+                    }
+                    break;
+                case 'partner':
+                    if (!$module->getOnlineInfo('isThirdParty')) {
```
Comment (luke-) [inline protected/humhub/modules/marketplace/Events.php:245]: Show this badge when `isPartner` is `true`. (New Field not avail yet)

File: protected/humhub/modules/marketplace/Events.php [hunk 5]
```diff
@@ -8,10 +8,19 @@
...
+use humhub\modules\marketplace\models\Module as ModelModule;
+use humhub\modules\ui\icon\widgets\Icon;
+use humhub\modules\ui\menu\MenuLink;
+use humhub\widgets\Button;
 use Yii;
 use yii\base\BaseObject;
 use yii\base\Event;
-use yii\helpers\Url;
 
 class Events extends BaseObject
```
Comment (luke-) [inline protected/humhub/modules/marketplace/Events.php:25]: In the whoile file we should check the `Module::$enabled` is `true` before handling the Events.

File: protected/humhub/components/ModuleManager.php [hunk 6]
```diff
@@ -270,6 +278,56 @@ public function getModules($options = [])
         return $modules;
     }
 
+    public function filterModules(array $modules, $filters = []): array
```
Comment (luke-) [inline protected/humhub/components/ModuleManager.php:281]: Some PHPDoc would be nice

File: protected/humhub/components/ModuleManager.php [hunk 7]
```diff
@@ -270,6 +278,56 @@ public function getModules($options = [])
...
+
+        $modules = $this->filterModulesByKeyword($modules, $filters['keyword']);
+
+        $modulesEvent = new ModulesEvent(['modules' => $modules]);
+        $this->trigger(static::EVENT_AFTER_FILTER_MODULES, $modulesEvent);
+
+        return $modulesEvent->modules;
+    }
+
+    public function filterModulesByKeyword(array $modules, $keyword = null): array
```
Comment (luke-) [inline protected/humhub/components/ModuleManager.php:295]: Some PHPDoc would be nice

**Review by luke- — CHANGES_REQUESTED** [review]
Thank you for the great work. In my first tests everything worked as expected. We still need to test & improve some details in the next week and Angela needs to check the design again.

I also tested without the "Marketplace" module and everything worked. Awesome!

Besides my PR notes, there are also following points:



- Enabled Modules should displayed before Disabled
- Filter: Multiple Tags or only one Active? Maybe we should add an option here. The `Tag` filter would probably used more often in the next time. 
- The `Marketplace::$enabled` variable has no effect

**Comment by yurabakhtin** [comment]
@luke- 

> Enabled Modules should displayed before Disabled

Commit https://github.com/humhub/humhub/pull/5476/commits/03e984d7ae5478b39d2e2d0f1bc7fdf1d24d1a06

**Comment by yurabakhtin** [comment]
@luke- 

> Filter: Multiple Tags or only one Active? Maybe we should add an option here. The Tag filter would probably used more often in the next time.

Sorry, not clear what to do here. Do you mean to create some new sub-option for the filter, it may be like a switcher between options:

- `AND` - find modules which have all selected tags
- `OR` - find modules which have at least one selected tag

or do you want to keep some option for the filter like:

- `'allowMultipleTags' => false` so it will be allowed only select one tag at the moment.
- `'allowMultipleTags' => true` allow to select multiple tags (as it is implemented currently)

?

**Comment by yurabakhtin** [comment]
@luke- 

> The `Marketplace::$enabled` variable has no effect

Commit https://github.com/humhub/humhub/pull/5476/commits/2098805a472c57b7d5b60913032743d0128f338a

File: protected/humhub/modules/admin/views/layouts/module-old.php (hunk 1 above)
Comment (yurabakhtin) [reply protected/humhub/modules/admin/views/layouts/module-old.php:1]: It is still used for url `/marketplace/purchase`

File: protected/humhub/modules/marketplace/controllers/PurchaseController.php (hunk 2 above)
Comment (yurabakhtin) [reply protected/humhub/modules/marketplace/controllers/PurchaseController.php:32]: Please see where it is used, maybe we should somehow move this into new lauyout:
purchase

File: protected/humhub/modules/marketplace/Events.php (hunk 3 above)
Comment (yurabakhtin) [reply protected/humhub/modules/marketplace/Events.php:255]: https://github.com/humhub/humhub/pull/5476/commits/f4163d88cac591e645cc04817f763a0bd42ca509#diff-d29d9ca029353610b8cbd988563fced151f48e64f7b74fe9341335aec477697bR256

File: protected/humhub/modules/marketplace/Events.php (hunk 4 above)
Comment (yurabakhtin) [reply protected/humhub/modules/marketplace/Events.php]: https://github.com/humhub/humhub/pull/5476/commits/f4163d88cac591e645cc04817f763a0bd42ca509#diff-d29d9ca029353610b8cbd988563fced151f48e64f7b74fe9341335aec477697bR261

File: protected/humhub/components/ModuleManager.php (hunk 6 above)
Comment (yurabakhtin) [reply protected/humhub/components/ModuleManager.php:288]: https://github.com/humhub/humhub/pull/5476/commits/f4163d88cac591e645cc04817f763a0bd42ca509#diff-46a3b644fa0848e9b9ff89b55e78b267182b69d39cea2f61870c54d37d0a78a5R281

File: protected/humhub/components/ModuleManager.php (hunk 7 above)
Comment (yurabakhtin) [reply protected/humhub/components/ModuleManager.php:309]: https://github.com/humhub/humhub/pull/5476/commits/f4163d88cac591e645cc04817f763a0bd42ca509#diff-46a3b644fa0848e9b9ff89b55e78b267182b69d39cea2f61870c54d37d0a78a5R302

File: protected/humhub/modules/marketplace/Events.php (hunk 5 above)
Comment (yurabakhtin) [reply protected/humhub/modules/marketplace/Events.php:25]: https://github.com/humhub/humhub/pull/5476/commits/910bb0c77beed80b6af022e185b1cc42feca7a9f

File: static/js/humhub/humhub.directory.js [hunk 8]
```diff
@@ -6,6 +6,36 @@ humhub.module('directory', function(module, require, $) {
         $(evt.$trigger).closest('form').submit();
```
[the remark this answers is not in the document]
Comment (yurabakhtin) [reply static/js/humhub/humhub.directory.js:6]: Commit https://github.com/humhub/humhub/pull/5476/commits/aab88298d6a150ac290c668566efec6c4a66c0c0

File: protected/humhub/modules/marketplace/controllers/PurchaseController.php (hunk 2 above)
Comment (luke-) [reply protected/humhub/modules/marketplace/controllers/PurchaseController.php:32]: @yurabakhtin What do you think of following approach?
image

File: protected/humhub/modules/admin/views/layouts/module-old.php (hunk 1 above)
Comment (luke-) [reply protected/humhub/modules/admin/views/layouts/module-old.php:1]: @yurabakhtin See here: https://github.com/humhub/humhub/pull/5476#discussion_r782399237

File: protected/humhub/components/Module.php [hunk 9]
```diff
@@ -32,6 +35,11 @@ class Module extends \yii\base\Module
      */
     private $_moduleInfo = null;
 
+    /**
+     * @var array the cached info loaded from online
+     */
+    private $_onlineInfo = null;
```
[the remark this answers is not in the document]
Comment (luke-) [reply protected/humhub/components/Module.php]: @yurabakhtin Hmm, I'm not a big fan of behaviors. Can you maybe separate it more clearly?

maybe something like e.g. 
```php
\humhub\modules\marketplace\components\OnlineModule

$onlineModule = new OnlineModule(Yii::$app->getModule('wiki'));
$categories = $onlineModule->getCategories();
```

[the remark this answers is not in the document]
Comment (yurabakhtin) [reply protected/humhub/components/Module.php]: @luke- Ok, commit https://github.com/humhub/humhub/pull/5476/commits/283056dd9ce4e09bb8625c4069505852848c2dc3.

File: protected/humhub/modules/marketplace/controllers/PurchaseController.php (hunk 2 above)
Comment (yurabakhtin) [reply protected/humhub/modules/marketplace/controllers/PurchaseController.php]: @luke- Done in commit https://github.com/humhub/humhub/pull/5476/commits/74ea0a6ea228b7db17f447810ed6e8ea1325cd9b

File: protected/humhub/modules/admin/views/layouts/module-old.php (hunk 1 above)
Comment (yurabakhtin) [reply protected/humhub/modules/admin/views/layouts/module-old.php]: Done.

**Review by luke- — APPROVED** [review]

**Review by luke- — APPROVED** [review]

**Comment by funkycram** [comment]
Great job, thanks!
Only one thing: I could not found the way to access to the purchased licence keys.
The modal box page exists, because if we add a new key to an uninstalled module, the purchased keys shows up.
Perhaps could we add a modal button to reach the modal page `/marketplace/purchase`?
The button could be named "Purchases" (same name as the former tab).
Would you like me to send a PR for that?

**Comment by yurabakhtin** [comment]
@funkycram Do you find this?:
keys

**Comment by funkycram** [comment]
Yes, but this is the only way to have this page with the purchased licence keys.
For example, if you have 2 modules purchased, the only way to see the 2 licence keys of these modules is to find another module and try to add a licence key:
image

**Comment by yurabakhtin** [comment]
@luke- 

> Perhaps could we add a modal button to reach the modal page /marketplace/purchase?
The button could be named "Purchases" (same name as the former tab).

Do you agree this?

**Comment by luke-** [comment]
I am not sure. It is of course a bit more hidden now, but you always get to the license key management via each paid module.
Or am I missing something here?

I would rather leave it out.

Would that be okay for you? Or am I overlooking use cases here?

**Comment by ArchBlood** [comment]
I'm also concerned about the missing admin sidebar section, was this intended or something? It currently doesn't make much sense to have to backtrack so much to get places.

Screenshot_1
Screenshot_2

**Comment by ArchBlood** [comment]
I've also noticed that not all modules are treated the same by the system, some show as "Partner" others don't, and if a module is a custom one that isn't within the marketplace then it shows as "Official" instead of no tag, whereas ones uploaded to the marketplace that isn't Official or Partner shows as no tags, I'm sure this was not meant and should be fixed.

Screenshot_3
Screenshot_4

**Comment by funkycram** [comment]
> I am not sure. It is of course a bit more hidden now, but you always get to the license key management via each paid module. Or am I missing something here?
> 
> I would rather leave it out.
> 
> Would that be okay for you? Or am I overlooking use cases here?

In this case, perhaps, for a purchased module, we could add an item in the menu:
image => image

Otherwise, with Humhub 1.11.0-beta.2, as the purchases tab is removed, you cannot find the licence keys of the purchased modules (unless you find another not installed module and try to add key).

Another, related, problem:
If I delete a licence key of a paid module (on https://partner.humhub.com/), then the customer cannot see that the licence key has been deleted and if later I send him a new key, he cannot add the new one. He need to uninstall the module to be able add the key again, or find another module to add the key which is quite confusing.

E.g. screenshot of a paid and installed module for which I have removed the licence key:
image

We should have the "Add licence key" even if the module is installed, if the key has been removed from https://partner.humhub.com
Thanks!

**Comment by luke-** [comment]
@funkycram Ok got it.

Would it be ok or sufficient for you if we:

- For all "Non-Free" modules, show "My License Keys" in the context menu. (Instead of nothing atm).
- Maybe for "Non-Free" modules with an added license, integrate a "$" as a symbol, so that it is clear that it was purchased.

---

Expired licenses could then be added again by "My Licecene"?

Unfortunately, there are currently no prominent notices about expired licenses. (At the moment, this is only displayed when an update is attempted.) Maybe we should open a separate issue for 1.12 to improve this?

**Comment by funkycram** [comment]
Thanks @luke- that would be fine!

For expired licenses, I'm not sure it's a problem not having notices about them, as they already receive 2 reminders emails (I mean for CUZY.APP modules) and, as you say, it's displayed when an updated is attempted, I think it's enough. If I have feedbacks from users on this subject I will forward them.

**Comment by luke-** [comment]
@yurabakhtin Can you please add this context option here for already purchased modules as well?

image

**Pull request merged**